ext/mbstring/mbstring.c

```diff
@@ -2941,6 +2941,130 @@ PHP_FUNCTION(mb_strtolower)
 	RETURN_STR(mbstring_convert_case(PHP_UNICODE_CASE_LOWER, ZSTR_VAL(str), ZSTR_LEN(str), enc));
 }
 
+static bool is_trim_wchar(uint32_t w, uint32_t *characters, size_t characters_len)
+{
+	for (size_t i = 0; i < characters_len; i++) {
+		if (w == characters[i]) {
+			return true;
+		}
+	}
+	return false;
+}
+
+static zend_string* trim_each_wchar(zend_string *str, zend_string *what, int mode, const mbfl_encoding *enc)
+{
+	unsigned char *in = (unsigned char*)ZSTR_VAL(str);
+	unsigned char *what_in = (unsigned char*)ZSTR_VAL(what);
+	uint32_t wchar_buf[128];
+	uint32_t what_wchar_buf[128];
+	size_t in_len = ZSTR_LEN(str);
+	size_t out_len = 0;
+	size_t what_out_len = 0;
+	size_t what_len = ZSTR_LEN(what);
+	unsigned int state = 0;
+	long left = 0;
+	long right = 0;
+	size_t total_len = 0;
+	bool first_left = true;
+
+	while (in_len) {
+		out_len = enc->to_wchar(&in, &in_len, wchar_buf, 128, &state);
+		ZEND_ASSERT(out_len <= 128);
+		total_len += out_len;
+
+		while (what_len) {
+			what_out_len = enc->to_wchar(&what_in, &what_len, what_wchar_buf, 128, &state);
+			ZEND_ASSERT(what_out_len <= 128);
+
+			if (first_left && (mode & 1)) {
+				for (size_t i = 0; i < out_len; i++) {
+					uint32_t w = wchar_buf[i];
+					if (is_trim_wchar(w, what_wchar_buf, what_out_len)) {
+						left += 1;
+					} else {
+						break;
+					}
+				}
+				first_left = false;
```

Comment:
I'm not sure I understand what this `first_left` is for. Let's say we have a string with more than 128 unicode whitspace characters at the left side, does this then work correctly?

Reply:
Yes, you're right. I'll fix this.
```sh
$ sapi/cli/php -r 'var_dump(mb_trim(str_repeat("　", 129)));'
string(3) "　"
```

Reply:
`mb_rtrim` is still has problem... I fixing this.
I wish I could read it from the right.

Reply:
This problem is fixed in f207f22c39f40493737c6deafeb5e481ee50b318.